### PR TITLE
Correctly handle money amounts

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
https://phabricator.wikimedia.org/T130525

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
@WMDE-Fisch @manicki Please beware of this class, it might be of use in the backend app as well

**Comment by JeroenDeDauw** [comment]
If anyone can think of a case that is not tested yet, please point it out. Kinda important this code does not subtly alter the values.

File: src/Domain/Model/Euro.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	 */
+	public static function newFromString( string $euroAmount ): self {
+		if ( !is_numeric( $euroAmount ) ) {
+			throw new InvalidArgumentException( 'Not a number' );
+		}
+
+		return new self( intval(
+			round( $euroAmount, self::$DECIMAL_COUNT ) * self::$CENTS_PER_EURO
+		) );
+	}
```
Comment (JeroenDeDauw) [inline src/Domain/Model/Euro.php]: Method loosely based on https://github.com/sebastianbergmann/money/blob/master/src/Money.php#L80

**Comment by JeroenDeDauw** [comment]
I don't like how this reads: `new Euro( 1000 )`. Looks like that is 1000 Euro, while it's just 10. Easy for someone to get confused. Perhaps we should make the constructor private and have `newFromCents( int $i )` and `newWholeAmount( int $i )` methods or something like that?

File: src/Domain/Model/Euro.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+use InvalidArgumentException;
+
+/**
+ * @licence GNU GPL v2+
+ * @author Jeroen De Dauw < jeroendedauw@gmail.com >
+ */
+final class Euro {
+
+	private static $DECIMAL_COUNT = 2;
+	private static $CENTS_PER_EURO = 100;
```
Comment (KaiNissen) [inline src/Domain/Model/Euro.php]: Is there a reason to use static members rather than constants here?

File: src/Domain/Model/Euro.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	private $cents;
+
+	/**
+	 * @param int $cents
+	 * @throws InvalidArgumentException
+	 */
+	public function __construct( int $cents ) {
+		if ( $cents < 0 ) {
+			throw new InvalidArgumentException( 'Amount needs to be positive' );
```
Comment (KaiNissen) [inline src/Domain/Model/Euro.php]: Not sure about this. Our use cases require amounts to be positive, in general a currency value might as well be negative. So isn't this something that should be handled by the validators?

File: src/Infrastructure/AmountParser.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * @licence GNU GPL v2+
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ * @author Jeroen De Dauw < jeroendedauw@gmail.com >
+ */
+class AmountParser {
+
+	private $locale;
+
+	public function __construct( string $locale = 'de_DE' ) {
```
Comment (KaiNissen) [inline src/Infrastructure/AmountParser.php]: Make `$locale` required maybe? There are no cases in which `AmountParser` is created without passing this parameter, anyway.

**Comment by KaiNissen** [comment]
I totally agree to your comment above regarding the constructor taking cents.

Nice production code/test code ratio, by the way! :)

File: src/Domain/Model/Euro.php (hunk 2 above)
Comment (JeroenDeDauw) [inline src/Domain/Model/Euro.php]: Yeah - they can be private. Once we switch to PHP 7.1, we can have private consts. Not that it matters much anyway

File: src/Domain/Model/Euro.php (hunk 3 above)
Comment (JeroenDeDauw) [inline src/Domain/Model/Euro.php]: This class is geared towards our usecases. It is not meant to be a general thing, which would be more complex and less safe for us.

**Comment by JeroenDeDauw** [comment]
> Nice production code/test code ratio, by the way! :)

I actually ran phploc on the files to get the LLOC. That is a lot more skewed than the LOC.

File: src/Infrastructure/AmountParser.php (hunk 4 above)
Comment (JeroenDeDauw) [inline src/Infrastructure/AmountParser.php]: Hmm, sure. Beware that it is you that added the optional thing here, I just moved it around

```php
public function setAmountFromString( string $value, string $locale = 'de_DE' ) {
```

**Pull request merged**